### PR TITLE
Initial implementation of Typescript server bindings

### PR DESCRIPTION
**Pull request by crisidev**
I'd like to merge this PR as soon as it looks good. This is not a final product, not really a working one, but it lays down the foundations for generating proper typescript bindings.

**Please note that the CI will fail on this PR until it is merged since the codegen diff cannot work properly until the project has been merged the first time.**

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: codegen-server-test/typescript/model/pokemon.smithy
Comment (pose) [inline codegen-server-test/typescript/model/pokemon.smithy:2]: (nitpick) typo `$econcile` -> `reconcile`

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsCodegenServerPlugin.kt
Comment (pose) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsCodegenServerPlugin.kt:29]: Python -> TypeScript

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsCodegenServerPlugin.kt
Comment (pose) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsCodegenServerPlugin.kt:53]: PythonServerCodegenVisitor -> TSServerCodegenVisitor

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsCodegenServerPlugin.kt
Comment (pose) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsCodegenServerPlugin.kt:72]: Does this even apply?

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCargoDependency.kt
Comment (pose) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCargoDependency.kt:22]: I believe we need to get rid of this one.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCargoDependency.kt
Comment (pose) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCargoDependency.kt:30]: What is this?

**Review by pose — CHANGES_REQUESTED** [review]

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCargoDependency.kt
Comment (82marbag) [reply codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCargoDependency.kt:30]: The list of dependencies that could be in your Cargo.toml, if used

**Comment by 82marbag** [comment]
TODO: add a README

File: CODEOWNERS
Comment (crisidev) [inline CODEOWNERS:16]: Hum, the group exists: https://github.com/orgs/awslabs/teams/smithy-rs-typescript-server/members

Do I have to do anything else here? 😕

Comment (unexge) [reply CODEOWNERS:16]: I think you need to add smithy-rs repo to the team: https://github.com/orgs/awslabs/teams/smithy-rs-typescript-server/repositories

**Review by unexge — APPROVED** [review]
I think we're still using `js` instead of `ts` in some places. It'd be nice if we clean those before merging.

**Comment by crisidev** [comment]
I'll double check everything tomorrow! Thanks for the review.

**Comment by pose** [comment]
This is related to: https://github.com/napi-rs/napi-rs/pull/1516

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt:114]: @pose this should go away, we will make this documentation based in the future.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/RustServerCodegenTsPlugin.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/RustServerCodegenTsPlugin.kt:90]: @pose please take a look to how this is done in Python now as the API changed

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCodegenVisitor.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/TsServerCodegenVisitor.kt:105]: @pose please take a look to how this is done in Python now as the API changed

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt
Comment (pose) [reply codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt:114]: Done (resolve button doesn't work yet).

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/RustServerCodegenTsPlugin.kt
Comment (pose) [reply codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/RustServerCodegenTsPlugin.kt:90]: Done. Seems like `model` was removed from almost everywhere.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt
Comment (pose) [reply codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt:126]: We are removing this file entirely.

File: codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/RustWriter.kt
Comment (crisidev) [inline codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/RustWriter.kt:471]: Ideally we should move this to a specific rust writer, but the amount of refactor to do it did not seems worth it.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customize/ServerCodegenDecorator.kt
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customize/ServerCodegenDecorator.kt:60]: This change is a no-op, shouldn't we just go back to what it was before?

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (unexge) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:213]: That should be fine, you just need to use reference I believe
 ```suggestion
                        #{SmithyServer}::socket::new_socket(address, port, backlog)
```

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:229]: This comment doesn't seems to be current.

Comment (unexge) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:229]: Yes, we're in `impl TsSocket`

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (unexge) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:265]: Same, you just need to use reference `#{SmithyServer}` instead of `aws_smithy_http_server`

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:213]:
```suggestion
                        #{SmithyServer}::socket::new_socket(address, port, backlog)
```

This works because you have a `rustTemplate` that interpolates over the `codegenScope` list.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (unexge) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:277]: Seems fine to me, is there a problem?

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:163]: This have to change to that we use the proper service name coming from the smithy model we are generating for instead of the hardcoding the PokemonService.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:176]: Let's remove commented code.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:265]: These comments are going into the generated code. Is the right place for them? On top of it, I don't really understand what this comment mean.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:287]:
```suggestion
                .serve(#{SmithyServer}::routing::IntoMakeService::new(app));
```

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:290]: This is already the better, non blocking way to spawn the server. I don't understand this comment.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:277]: Let's remove this comment, it doesn't produce value.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:210]: I don't understand the comment. All the comments labelled with `XXX` seems more notes for yourself than actual comments. Can we remove all of them since they will end up in the code generated code?

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsServerOperationErrorGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsServerOperationErrorGenerator.kt:39]: I would move this comment on the top of the `renderFromTsErr` function so that it does not end up in the generated code.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsServerOperationErrorGenerator.kt
Comment (crisidev) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsServerOperationErrorGenerator.kt:20]: We do not depend anymore from `ServerOperationErrorGenerator`

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (unexge) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:177]: We're not exporting anything here as opposed to `TsSocket`, is there a reason to not use `#{SmithyTs}::server::start_hyper_worker(socket, service).expect("failed to start the hyper server");`?

File: rust-runtime/aws-smithy-http-server-typescript/examples/pokemon-service.ts
Comment (unexge) [inline rust-runtime/aws-smithy-http-server-typescript/examples/pokemon-service.ts:75]: Do we need to expose `TsSocket` here? Can't we just handle that in `app.start('127.0.0.1', 9090)`?

File: rust-runtime/aws-smithy-http-server-typescript/src/lib.rs
Comment (crisidev) [inline rust-runtime/aws-smithy-http-server-typescript/src/lib.rs:6]: Both modules here are not used because of the issue with napi not being able to re-export from another crate. 

I would completely get rid of this crate for now, since it is unused and rely on code generation until there is a solution where we can abstract this code into a runtime crate.

**Review by pose — APPROVED** [review]

**Review by unexge — APPROVED** [review]

File: codegen-server/typescript/src/main/resources/META-INF/services/software.amazon.smithy.build.SmithyBuildPlugin
Comment (jjant) [inline codegen-server/typescript/src/main/resources/META-INF/services/software.amazon.smithy.build.SmithyBuildPlugin:1]: What does this file do?

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (jjant) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:263]: Looks like you could combine most of this into a single `rustTemplate` which would be a lot more readable.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (jjant) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:278]: Prefer single `rustTemplate` over splitting them like this.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt
Comment (jjant) [inline codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt:97]: I think you can write your own `writeJson` function with a `@Language` annotation so that you get syntax highlighting on the json, example: https://github.com/awslabs/smithy-rs/blob/c7c82855986673fc950c532d4cbdfa6cee4cc201/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/RustWriter.kt#L146-L151

**Review by jjant — APPROVED** [review]
Only gave this PR a cursory look, but it looks good.

I've left some minor suggestions that are not blockers.

File: settings.gradle.kts
Comment (pose) [reply settings.gradle.kts:13]: We will do it for Python as well but not as part of this PR.

File: codegen-server/typescript/src/main/resources/META-INF/services/software.amazon.smithy.build.SmithyBuildPlugin
Comment (pose) [reply codegen-server/typescript/src/main/resources/META-INF/services/software.amazon.smithy.build.SmithyBuildPlugin:1]: It's required so it's discovered by the Smithy CLI (Thanks @unexge )

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (pose) [reply codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:263]: I rather keep it this way. I find it clearer.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt
Comment (pose) [reply codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt:278]: I think I'd rather leave it this way for now.

File: codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt
Comment (pose) [reply codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/customizations/TsServerCodegenDecorator.kt:97]: We would leave it like this for now (see comments above).

**Comment by pose** [comment]
**Heads up!** CI failures: The diff tool just checkouts to current main and tries to run `./gradlew codegen-server-test:typescript:assemble`. However, since the command does not exist yet in main, it fails. We will merge the change even if that step fails.

**Review by rcoh — APPROVED** [review]

**Pull request merged**
